component/receiver.go

```diff
@@ -105,6 +105,10 @@ type ReceiverCreateParams struct {
 	// Logger that the factory can use during creation and can pass to the created
 	// component to be used later as well.
 	Logger *zap.Logger
+	// SharedPerConfig is a user-provided value that gets shared between
+	// receiver creators that have the same config, if applied by the Receiver
+	// Factory.
+	SharedPerConfig interface{}
```

Comment:
Not sure I understand why the CreateParams needs to change. They represents parameters passed from the service to the components, and don't see this as one of the thing that the service needs to share.

Reply:
The main alternative I see is to make a bunch of new factory options WithTracesShared/WithLogsShared/WithMetricsShared that accept functions with a different signature that allows the shared instance to be injected directly as a function argument.  That just seemed a bit verbose.  Would you prefer that or do you have any other ideas of how to do it?

Reply:
> params is already a generic "catch-all" struct where additional stuff can be added in a backwards-compatible way. 

Indeed, but not for internal details like this.

An alternative is to define an "CreateTracesSettings" (in the helper) that include this and any other thing that we may pass to the `CreateTraceReceiver` in the future, and include SharedPerConfig:

```golang
CreateTracesSettings {
	params component.ReceiverCreateParams,
	cfg configmodels.Receiver,
	consumer consumer.TraceConsumer,
        SharedPerConfig interface{}
}

type CreateTraceReceiver func(context.Context, CreateTracesSettings) (component.TraceReceiver, error)

```

Reply:
@bogdandrutu alright I changed it to just use another set of CreateShared*Receiver options to keep things backwards compatible (there were a ton of things already using the existing option signatures).